nixos/modules/services/networking/infiniband.nix

```diff
@@ -0,0 +1,78 @@
+{lib, pkgs, config, ...}:
+
+let
+  cfg = config.networking.ib-interfaces;
+in {
+
+  imports = [];
+
+  options = {
+    networking.ib-interfaces = {
+      enable = lib.mkEnableOption "ib-interfaces";
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+
+    environment.systemPackages = [ pkgs.rdma-core ];
+    services.udev.packages =
+      let patchedUdevRules = pkgs.stdenv.mkDerivation {
+        name = "patched-${pkgs.rdma-core.name}-udev-rules";
+        src = "${pkgs.rdma-core}/lib/udev/rules.d";
+        phases = ["unpackPhase" "patchPhase" "installPhase"];
+        patchPhase = ''
+          substituteInPlace 60-srp_daemon.rules \
+            --replace /bin/systemctl ${pkgs.systemd}/bin/systemctl
+        '';
+        installPhase = ''
+          mkdir -p $out/lib/udev/rules.d
+          cp -r . $out/lib/udev/rules.d/
+        '';
+      }; in [ patchedUdevRules ];
+
+    # Units
+    systemd.units = with pkgs; {
+      "ibacm.service".text =
+        builtins.readFile "${rdma-core}/lib/systemd/system/ibacm.service";
```

Comment:
This wont work in restricted mode and probably also reads outside of the sandbox. Since `readFile` isn't really "pure" as it is used here.

I would recommend going for something like I sketched below for all your units. That removes the need for `readFile` and also is a lot shorter / more readable(?).

```
let
  mkUnit = name: lib.nameValuePair name ({
    unit = pkgs.runCommand "unit" { preferLocalBuild = true; } ''
      mkdir -p $out
      ln -s ${rdma-core}/lib/systemd/system/${name} $out/${name}
    '';
  });
in {
  systemd.units = lib.listToAttrs (map mkUnit [
    "ibacm.socket"
    "iwpmd.service"
    "rdma-hw.target"
    "rdma-load-modules@.service"
    "rdma-ndd.service"
    "srp_daemon_port@.service"
    "srp_daemon.service"
  ]);
}
```